Automattic/wp-calypso
# Explat: Implement experiment dispatcher

## Proposed Changes

This implements a neat and reusable dispatcher for ExPlat experiments. By dispatcher, I mean a tool that will redirect the user to different URLs based on their assignment.

To define an experiment, all you need to do is add something like the following object:

```ts
{
	signup: {
		title: 'Sign Up Experiment - Between /start and /setup',
		description: 'This experiment is to test the sign up flow between /start and /setup',
		experiment_explat_id: 'calypso_signup_onboarding_aa_test',
		variants: {
			control: {
				description: 'The current sign up flow',
				url: '/start',
			},
			treatment: {
				description: 'The new sign up flow',
				url: '/setup/onboarding',
			},
		},
	}
}
```
Then it will automatically retrieve user authentication status, fetch the experiment, see the matching branch, then redirect the user accordingly. 

## Why are these changes being made?
To make comparing two /setup and /start more equal. Before this, we sent the user to /start anyways, and within start we hard-redirected them to /setup if they fall into treatment. But if they fall into `control`, the were not redirected. Giving /start an edge. Now both branches are redirected.

## Testing Instructions

1. Go to 22136-explat-experiment and assign yourself to a branch. 
1. Go to /start/assign/signup.
2. You should be redirected to /start if you assign control, and to /setup/onboarding if you assign treatment.

## Next step
- We should ping the system to whitelist /assign/* and point it towards Calypso. 
- The LOHP should point to /assign/signup.
